### DISCUSSION TITLE
DataTable column content alignmnet

### DISCUSSION THREAD

**Discussion by luked42** [Textualize/textual#4463] [Q&A]
Hey, 

I'm finding that when adding rows to a DataTable after clearing and readding rows that the alignment of content in the column 'snaps' to the correct position after the row is highlighted. I'm trying to understand why my DataTable is behaving like this and what can be done to fix the issue.

```py
from textual.app import App, ComposeResult
from textual.widgets import DataTable

class MainWindow(App):
    def compose(self) -> ComposeResult:
        self.ideas = ["a"]
        yield DataTable()

    def on_mount(self) -> None:
        table = self.query_one(DataTable)
        table.add_columns(*('foo', 'bar'))

        self.redraw()

    def on_data_table_row_selected(self, event: DataTable.RowSelected) -> None:
        self.ideas.append(f"{self.ideas[-1] * 2}")
        self.redraw()

    def redraw(self) -> None:
        table = self.query_one(DataTable)
        table.clear()
        for data in self.ideas:
            table.add_row(*(data, '-'), key=str(data))
        table.cursor_type = 'row'

if __name__ == "__main__":
    app = MainWindow()
    app.run()
```

https://github.com/Textualize/textual/assets/5997885/aed8a7d7-65fd-46e9-a8d2-de2c7160400f

Some things I note from this
* When I move the cursor row onto a new row, that row expands to where it should align with
* When I add a new row, the previously existing rows align with one another, and the header row aligns as well
* When I move the cursor row, the header row won't align with the newly added row.

Is this a common issue people work through? 

Something to note: 
If instead of calling `redraw`  / `clear` every time I add a row I append a row, I don't run into this issue.

Thanks for any insights you can give, love the framework!

**Comment by TomJGooding** [comment]
I suspect this is related to this issue about when the column widths are updated:

- #3449

**Comment by TomJGooding** [comment]
After a bit more experimentation with this, actually I think there might be something else going on.

For reasons I don't understand yet, this column width issue only seems to occur when redrawing the table when handling the selected event. Here's an example instead using a binding, which seems to work as expected. 

```python
from textual.app import App, ComposeResult
from textual.widgets import DataTable


class MainWindow(App):
    BINDINGS = [("space", "update_table")]

    def compose(self) -> ComposeResult:
        self.ideas = ["a"]
        yield DataTable()

    def on_mount(self) -> None:
        table = self.query_one(DataTable)
        table.add_columns(*("foo", "bar"))
        table.cursor_type = "row"

        self.redraw()

    # def on_data_table_row_selected(self, event: DataTable.RowSelected) -> None:
    def action_update_table(self) -> None:
        self.ideas.append(f"{self.ideas[-1] * 2}")
        self.redraw()

    def redraw(self) -> None:
        table = self.query_one(DataTable)
        table.clear()
        for data in self.ideas:
            table.add_row(*(data, "-"), key=str(data))


if __name__ == "__main__":
    app = MainWindow()
    app.run()
```

**Reply by luked42** [reply]
Hey Tom, thanks for your response. I've replied in the now issue at https://github.com/Textualize/textual/issues/4470#issuecomment-2116042162
